### ISSUE TITLE
UnnecessaryLet: false positive when variable used directly

### ISSUE THREAD

**Issue by osipxd** [detekt/detekt#4248]
## Expected Behavior

```kotlin
token?.let { addHeader("X-Access-Token", it) }    // Compilant, it's ok
token?.let { addHeader("X-Access-Token", token) } // Should be also compliant, but it isn't
```

Variable usage inside `let` block is smart-casted to non-nullable and `let` call is not unnecessary.

## Your Environment

* Version of detekt used: v1.18.0

**Comment by BraisGabin** [comment]
I would say that in this case the fisrt one is a false negative instead of the second one beeing a false positive. Both cases can be replaced by an `if` statment:

```kotlin
if (token != null) addHeader("X-Access-Token", token)
```

**Comment by osipxd** [comment]
Hm, according to examples from [docs](https://detekt.github.io/detekt/style.html#unnecessarylet) it is considered as compliant code 🤔 
But I agree that `if` will look cleaner

**Comment by BraisGabin** [comment]
I think that the docs are a bit missleading and we should improve them.

Compilant:
```kotlin
class A {
  var token: String? = null

  fun foo() {
    token?.let { println(it) } // Here the if is not possible because between the check and the usage the value could change
  }
}
```

Non-compilant:
```kotlin
fun foo() {
  val token: String? = getToken()
  token?.let { println(it) } // Here it is safe to change the let for an if
}
```

In our samples we just have `a?.let { print(it) }` and that's an over simplification.

**Comment by BraisGabin** [comment]
My proposal would be:

```diff
  * <noncompliant>
  * a.let { print(it) } // can be replaced with `print(a)`
  * a.let { it.plus(1) } // can be replaced with `a.plus(1)`
  * a?.let { it.plus(1) } // can be replaced with `a?.plus(1)`
  * a?.let { that -> that.plus(1) }?.let { it.plus(1) } // can be replaced with `a?.plus(1)?.plus(1)`
  * a.let { 1.plus(1) } // can be replaced with `1.plus(1)`
  * a?.let { 1.plus(1) } // can be replaced with `if (a != null) 1.plus(1)`
+ * fun foo() {
+ *   val a: Any? = getA()
+ *   a?.let { print(it) } // can be replaced by `if (a != null) println(a)`
+ * }
+ * fun foo() {
+ *   val a: Any? = getA()
+ *   a?.let { print(a) } // can be replaced by `if (a != null) println(a)`
+ * }
  * </noncompliant>
  *
  * <compliant>
- * a?.let { print(it) }
+ * class A {
+ *   var a: String? = null
+ *
+ *   fun foo() {
+ *     a?.let { print(it) }
+ *   }
+ * }
+ * getA()?.let { print(it) }
  * a?.let { 1.plus(it) } ?.let { msg -> print(msg) }
  * a?.let { it.plus(it) }
  * val b = a?.let { 1.plus(1) }
  * </compliant>
```

I think that the key to know if `let` is unnecesary or not is basically if it is wokring over a local variable or if it's not. If it works over a local variable nearly always it is a false positive. If it works over a property or it works over an expresion (for example a function) it's ok to use it because the only way to avoid it is to use a local variable.

**Comment by vladimirfx** [comment]
I agree that for side effects calls (like `println()` and similar) let is usable only for nullable fields.

Another valid use case for `let()` is some form of `map()` for any receiver. Example from production code, 4 lines of similar code:
```kotlin
mapOf(
...
"appTimeStamp" to failure["timestamp"]?.let { Timestamp("$it".toLong()) },
...
)
```
Sure, this can be rewritten as if expression but this form is more readable.

**Comment by BraisGabin** [comment]
This case seems like a valid use to me. You are using the output of let and it is safe call.

**Comment by BraisGabin** [comment]
I'm not sure what to do with this. I have a PR with the needed unit tests to fix this PR. But I don't know where I can get the smart cast information that we need to implement it.

**Comment by chao2zhang** [comment]
`SMARTCAST` does exist in BindingContext though. [Source code](https://cs.github.com/JetBrains/kotlin/blob/33c122d231be31d1429282219685d773efcd0c50/compiler/frontend/src/org/jetbrains/kotlin/resolve/BindingContext.java?q=BindingContext#L163)
